HPCC-29981 Improve gateway support in ESDL script

timothyklemm opened the pull request:
- Add support for gateway URLs that are references to secrets that can be resolved by the query target.
- Add support for gateway URLs that are references to secrets that can be resolved by the ESP. Only url, username, and password secret properties are used to assemble a replacement URL for the gateway. Each secret must explicitly enable compatibility with this use case.
- Add support for explicit pass-through values not interpreted by the ESP.
- Ensure inline URLs can be updated with separately configured username and password values. These updates must be enabled by defining either Gateways/@resolveInlineURLs as true or, for backward compatibility, Gateways/@legacyTransformTarget as a non-empty value.

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-29981 Improve gateway support in ESDL script

- Add support for gateway URLs that are references to secrets that can be
  resolved by the query target.
- Add support for gateway URLs that are references to secrets that can be
  resolved by the ESP. Only url, username, and password secret properties
  are used to assemble a replacement URL for the gateway. Each secret must
  explicitly enable compatibility with this use case.
- Add support for explicit pass-through values not interpreted by the ESP.
- Ensure inline URLs can be updated with separately configured username and
  password values. These updates must be enabled by defining either
  Gateways/@resolveInlineURLs as true or, for backward compatibility,
  Gateways/@legacyTransformTarget as a non-empty value.

Signed-off-by: Tim Klemm <tim.klemm@lexisnexisrisk.com>

asselitx commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
+
+void EsdlServiceImpl::resolveGatewayInlineURL(IPTree& gateway) const
+{
+    const char* name = gateway.queryProp("@name");
```
Is `@name` guaranteed?

asselitx reviewed (changes requested):
I think there may be a couple small changes needed.

timothyklemm commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
+    const char* name = gateway.queryProp("@name");
+    StringBuffer url(gateway.queryProp("@url"));
+    if (url.isEmpty())
+        throw makeStringExceptionV(-1, "gateway %s: missing url property", name);
```
You commented on @name twice. I'll address here. There is a logical requirement for @name. In the legacy transform, @name becomes @ServiceName (see lines 1477 and 1479). There is a need for something that the roxie can use to select the gateway and we've assumed for years it is called @name.

That's not to say that nothing should change. What should the absence of @name mean? We've been assuming its presence for legacy transforms. Absence here, and at 1414, only affects error reporting. Should we default to something like "unknown" for error reporting, because we don't know there isn't another naming convention in use, but fail when we try to act on it?

asselitx replied:
I was only thinking in terms of the local effect of calling makeStringExceptionV with a  possible null pointer. If it seems overkill to have more checks then you should leave as-is.

timothyklemm commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
+        throw makeStringExceptionV(-1, "gateway %s: '%s' does not identify an 'esp' category secret", name, identification);
+    if (!isEmptyString(permission) && !secret->getPropBool(permission))
+        throw makeStringExceptionV(-1, "gateway %s: '%s' does not grant '%s' permission", name, identification, permission);
+    if (!secret->getProp("url", url.clear()) || url.isEmpty())
```
The url will always get cleared to prevent getProp from appending the secret's defined URL to the configurations secret identification . The second clause is a sanity check that the secret didn't define an empty URL. I think the logic is as intended.

asselitx replied:
Ok I guess I misunderstood. I thought there could be a valid case here where it wasn't required to resolve the secret.

asselitx commented on esp/services/esdl_svc_engine/esdl_binding.hpp:
```diff
@@ -210,11 +210,109 @@ class EsdlServiceImpl : public CInterface, implements IEspService
+    static constexpr const char* gwLocalSecretPrefix = "local-secret:";
+    static constexpr const size_t gwLocalSecretPrefixLength = strlen(gwLocalSecretPrefix);
+    static constexpr const char* gwPassThroughPrefix = "pass-through:";
+    static constexpr const size_t gwPassThroughPrefixLength = strlen(gwPassThroughPrefix);
```
Ok, I was thinking just for our developer ESDL scripting docs, but either way.

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
+        url.clear();
+        url.append(scheme);
+        url.append("://");
+        if (userinfo.length()>0 )
```
Extra space.  I also think the >0 on many of these ifs is a bit superfluous.  if (userinfo.length()) would suffice.

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
+            url.appendf("%s@", userinfo.str());
+        url.append(host);
+        if (port.length() > 0)
+            url.appendf(":%s", port.str());
```
url.append(':').append(port);
might be slightly faster.

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
+        throw makeStringExceptionV(-1, "gateway %s: '%s' is not of the form \"[ vault-id ':' ] secret-name\"", name, identification);
+    }
+    if (!secret)
+        throw makeStringExceptionV(-1, "gateway %s: '%s' does not identify an 'esp' category secret", name, identification);
```
Wording might be slightly confusing.  Mostly in the sense that I would say the secret wasn't found rather than "doesn't identify".  But the rest including mentioning the category is good.

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
+        throw makeStringExceptionV(-1, "gateway %s: missing url property", name);
+    if (strncmp(url, gwLocalSecretPrefix, gwLocalSecretPrefixLength))
+        throw makeStringExceptionV(-1, "gateway %s: expected '%s...'; got '%s'", name, gwLocalSecretPrefix, url.str());
+    const char* identification = url.str() + gwLocalSecretPrefixLength;
```
For ECL I currently require connection secrets to start with "http-connect-".  We should consider doing the same here.  It means the secrets were explictly created for this purpose.  I may regret it in the future, who knows, but it gives a little control and forced standardization.

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
+        tgtcfg.setown(createPTreeFromIPT(tgtcfg));
+        const char* qt = tgtcfg->queryProp("@querytype");
+        bool published = (qt && (strieq(qt, "roxie") || strieq(qt, "wsecl")));
+        const char* permission = (published ? "allowPublishedGatewayUsage" : "allowUnpublishedGatewayUsage");
```
Just because permission is used in so many overloaded ways I wonder if something like the term "requiredUsage" would work instead?

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
+    if (!secret->getProp("url", url.clear()) || url.isEmpty())
+        throw makeStringExceptionV(-1, "gateway %s: '%s' missing required 'url' property; credential-only secrets not supported", name, identification);
+    const char* username = secret->queryProp("username");
+    if (isEmptyString(username) && !secret->getPropBool("insecure"))
```
The term insecure is a bit ambiguous.  It makes me think of things like do we accept self signed certs or not verify the server.. some of those types of options may be needed for dev secrets in the future.
Can we call it something else?

afishbeck commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
+    gateway.setProp("@url", url);
+}
+
+void EsdlServiceImpl::adjustURL(StringBuffer& url, const char* username, const char* password) const
```
Can we put something in the method name about adding credentials?  adjustURL is a bit ambiguous.

afishbeck reviewed (commented):
@timothyklemm generally looks fine.  Some comments about naming, syntax, etc.

Commit:
Review feedback

Commit:
Refactor

- Anticipate alternate update requirements, such as for database access.
- Prevent two versions of one secret from being used in one transaction.

Commit:
additional cleanup

Commit:
more cleanup

Commit:
HPCC-26029 Improve ESDL script secret support

- Add a transactional secret cache to the script context.
- Add secret support to http-post-xml.
- Update secret support in mysql.

Signed-off-by: Tim Klemm <tim.klemm@lexisnexisrisk.com>

Commit:
Updates from review and discussion comments

Commit:
Retrofit to use IEsdlScriptContext secrets

Commit:
Secrets cache abstraction
